### PR TITLE
Add colours to output normal and refactored code

### PR DESCRIPTION
**Pull request by yokurang**

### CODE REVIEW

**Comment by yokurang** [comment]
Here I wanted to add refactoring in this PR because I also wanted to include the new functions for normal mode when refactoring

**Comment by yokurang** [comment]
I am also running into this particular error when trying to generate the `mli` files:

```
❯ 
❯ ocamlc -c -i lib/InteractiveViewer.ml
File "lib/InteractiveViewer.ml", line 1, characters 5-11:
1 | open Nottui
         ^^^^^^
Error: Unbound module Nottui
```

Was wondering if anyone might know how to resolve this? I made sure the `dune` file has `nottui` and that I have the package installed using `opam`.

**Review by panglesd — COMMENTED** [review]
PascalCasing is not very idiomatic in OCaml... But refactoring in multiple module is a very good idea!

A few comments below.

File: lib/interactive_viewer.ml [hunk 1]
```diff
@@ -104,6 +98,68 @@ let change_summary z_patches : ui Lwd.t =
...
+  let their_line_num = ref hunk.Patch.their_start in
+  let mine_line_num = ref hunk.Patch.mine_start in
+  let lines_ui =
+    List.map
+      (function
+        | `Common line ->
+            incr their_line_num;
+            incr mine_line_num;
+            W.string ~attr:Notty.A.empty
+              (Printf.sprintf "%2d %2d   %s" !mine_line_num !their_line_num line)
```
[the remark this answers is not in the document]
Comment (yokurang) [reply lib/interactive_viewer.ml]: Created new issue: https://github.com/panglesd/diffcessible/issues/66

File: dune-project [hunk 2]
```diff
@@ -7,9 +7,9 @@
 (source
  (github panglesd/diffcessible))
 
-(authors "Paul-Elliot <peada@free.fr>" "Jules Aguillon <jules@j3s.fr>")
+(authors "Paul-Elliot <peada@free.fr>" "Jules Aguillon <jules@j3s.fr>" "Alan Matthew <alan.matthew.yk@gmail.com>")
```
Comment (Julow) [inline dune-project:10]: Thanks for this!

**Review by Julow — COMMENTED** [review]
- I agree with the `.mli`. These modules have a complex implementation, which should be hidden behind a documented and abstracted interface. 
- The casing feels non-idiomatic for OCaml but why not. I won't enforce my favorite casing.

File: lib/HelpView.mli [hunk 3]
```diff
@@ -0,0 +1,13 @@
+(** This module provides UI components and attributes for a help panel in a Nottui-based application. *)
+
+val help_visible : bool Lwd.var
+(** [help_visible] is a variable indicating whether the help panel is currently visible. *)
+
+val help_panel : Nottui.ui
+(** [help_panel] constructs the user interface for the help panel. This includes various help-related commands and their corresponding visual attributes, making it easy for users to understand available interactions. *)
+
+val help_keyboard_area : Nottui.ui
+(** [help_keyboard_area] provides a keyboard interaction area where pressing 'q' will close the help panel. *)
```
Comment (panglesd) [inline lib/HelpView.mli:10]: This value is not used.

The problem would not be solved by simply removing the value. In my opinion, the problem is more that the logic for opening/closing the help panel is spread between two very different location: `HelpView.ml` and `InteractiveViewer.ml`. This logic should be in one place!

File: lib/InteractiveViewer.mli [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    @return Executes indefinitely, handling UI events until termination conditions are met (e.g., user quits). *)
+
+val start_test : Patch.t list -> char list -> int -> int -> unit
+(** [start_test patch events width height] facilitates automated UI testing by simulating keypress events and rendering the UI to the terminal.
+    This function is useful for automated or manual testing of UI behaviors in response to specific inputs.
+    @param patch A list of patches to be displayed in the test.
+    @param events A list of characters representing simulated keypresses to interact with the UI.
+    @param width The width of the rendering area in characters.
+    @param height The height of the rendering area in characters.
+    @return Outputs the resulting UI state to the terminal, providing visual feedback on the UI's response to the simulated events. *)
```
Comment (panglesd) [inline lib/InteractiveViewer.mli:21]: The return ocamldoc tag (`@return`) is usually used to explain the returned value of a function. In this case, the return value is unit.
The explanation for the side effect (outputing the UI state to the terminal) could be done outside of the tag. But I'm still fine with it like that!

File: lib/InteractiveViewer.mli [hunk 5]
```diff
@@ -0,0 +1,21 @@
+(** This module defines the main user interface and interaction logic for a diff viewer application using the Nottui library. It provides mechanisms to navigate through patches and view them in different formats, respond to user inputs, and run test simulations. *)
+
+val view : Patch.t list -> Nottui.ui Lwd.t
+(** [view patches] constructs a reactive UI for navigating and viewing a list of patches.
+    It supports toggling between help view and normal view, and handling navigation commands.
+    @param patches A list of patches to be displayed and navigated through.
+    @return A reactive UI element constructed using Lwd combinators. *)
+
```
Comment (panglesd) [inline lib/InteractiveViewer.mli:8]: This function is internal to this module and is actually an implementation detail. It should not be part of the interface.

For instance, suppose we decide to change the internal implementation of diffcessible and stop using `Nottui`. In this case, this function would change type, making it an unnecessary breaking change.

File: lib/InteractiveViewer.mli [hunk 6]
```diff
@@ -0,0 +1,21 @@
+(** This module defines the main user interface and interaction logic for a diff viewer application using the Nottui library. It provides mechanisms to navigate through patches and view them in different formats, respond to user inputs, and run test simulations. *)
```
Comment (panglesd) [inline lib/InteractiveViewer.mli:1]: "Using the nottui library" is an implementation detail, irrelevant to the user here! ;)

File: lib/PatchNavigation.mli [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+type direction =
+  | Prev
+  | Next
+      (** Type representing navigation directions within a list of patches. 
+          - [Prev] indicates navigation to the previous item.
+          - [Next] indicates navigation to the next item. *)
+
+val quit : bool Lwd.var
+(** [quit] is a variable indicating whether the application should terminate its main loop. *)
```
Comment (panglesd) [inline lib/PatchNavigation.mli:11]: Shouldn't this be in InteractiveViewer?

It seems to be unrelated to patch navigation to me!

File: lib/PatchNavigation.mli [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+val navigate : 'a Zipper.t Lwd.var -> direction -> unit
+(** [navigate z_patches dir] navigates within a zipper of patches in the specified direction [dir].
+    It updates the zipper's state to the previous or next item based on the direction.
+    @param z_patches A reactive variable containing a zipper of patches.
+    @param dir The direction to navigate within the zipper.
+    @return () Updates the state of [z_patches] without returning any value. *)
+
+val change_summary : Patch.t Zipper.t Lwd.var -> Nottui.ui Lwd.t
```
Comment (panglesd) [inline lib/PatchNavigation.mli:20]: This does not seem to be related with patch navigation. Either it should be moved, or the name of this module should be more general!

Alsto, this should not take a `'a Zipper.t Lwd.t`: the change summary has no reason to modify the variable!
I think the best signature for this would be: `Patch.t Zipper.t -> Nottui.ui`

File: lib/OperationView.mli [hunk 9]
```diff
@@ -0,0 +1,13 @@
+(** This module defines UI components for displaying information about operations in a diff viewer using the Nottui library. These components visualize different types of patch operations like creation, deletion, and edits. *)
+
+val operation_info : Patch.t Zipper.t Lwd.var -> Nottui.ui Lwd.t
```
Comment (panglesd) [inline lib/OperationView.mli:3]: This should have type: `Patch.t Zipper.t -> Nottui.ui`.

File: lib/OperationView.mli [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+(** [operation_info z_patches] creates a reactive UI element that displays information about the current operation in a zipper of patches.
+    It shows the current operation index, the total number of operations, and the number of hunks in the current patch.
+    @param z_patches A reactive variable containing a zipper of patches.
+    @return A reactive UI element that displays operation-related information. *)
+
+val current_operation : Patch.t Zipper.t Lwd.var -> Nottui.ui Lwd.t
```
Comment (panglesd) [inline lib/OperationView.mli:9]: This should have type `Patch.t Zipper.t -> Nottui.ui`

File: lib/HunkView.mli [hunk 11]
```diff
@@ -0,0 +1,23 @@
+(** This module provides functionalities for rendering different views of hunks within a patch in a diff viewer application using the Nottui library. It supports both unified and side-by-side display of changes. *)
+
+val current_hunks : Patch.t Zipper.t Lwd.var -> Nottui.ui Lwd.t
```
Comment (panglesd) [inline lib/HunkView.mli:3]: Here again, the Lwd parts of this function should be removed!

File: lib/HunkView.mli [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+(** [current_hunks z_patches] creates a reactive UI element to display all hunks in the current patch from a zipper of patches in a unified format.
+    @param z_patches A reactive variable containing a zipper of patches.
+    @return A reactive UI element listing all hunks in the current patch. *)
+
+val current_hunks_side_by_side : Patch.t Zipper.t Lwd.var -> Nottui.ui Lwd.t
```
Comment (panglesd) [inline lib/HunkView.mli:8]: Here again, the Lwd parts of this function should be removed!

File: lib/HunkView.mli [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    This view aligns changes horizontally, allowing easier comparison of 'mine' and 'their' versions.
+    @param z_patches A reactive variable containing a zipper of patches.
+    @return A reactive UI element listing all hunks in the current patch in side-by-side view. *)
+
+type view_mode =
+  | SideBySide
+  | Normal
+      (** Defines available view modes for displaying diffs: either side-by-side or as a unified view. *)
+
+val view_mode : view_mode Lwd.var
```
Comment (panglesd) [inline lib/HunkView.mli:19]: The logic for toggling view mode should either be all in this module, or all in another module, but not spread in multiple files!

File: lib/HunkView.ml [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+            updated_their_acc)
+
+let lines_with_numbers lines attr_change prefix =
+  let line_num = ref 0 in
+  List.fold_left
+    (fun acc line ->
+      match line with
+      | Common s ->
+          incr line_num;
+          let content = Printf.sprintf "%3d   %s" !line_num s in
```
Comment (panglesd) [inline lib/HunkView.ml:128]: Now all line numbers are considered to have 3 digits: that's a lot! :)

But no need to change anything in this PR: you have opened #66 anyway.

File: lib/HunkView.ml [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+      | `Common line :: t ->
+          let common_mine_acc = Common line :: updated_mine_acc in
+          let common_their_acc = Common line :: updated_their_acc in
+          split_and_align_hunk t common_mine_acc common_their_acc
+      | _ ->
+          split_and_align_hunk remaining_hunks updated_mine_acc
+            updated_their_acc)
+
+let lines_with_numbers lines attr_change prefix =
+  let line_num = ref 0 in
```
Comment (panglesd) [inline lib/HunkView.ml:122]: Let's not use a reference here!

**Review by panglesd — COMMENTED** [review]
Looking good!

I made a series of comments. Refactoring in separate modules is not an easy task! We can go through them in a pair-programming session.

A minor comment that I cannot make inline is also that `lib/zipper.ml` (and maybe `bin/main.ml`) should follow the same naming convention as other files (so you should capitalize it).

File: lib/HelpView.mli (hunk 3 above)
Comment (yokurang) [reply lib/HelpView.mli:10]: Move help_visible variable into InteractiveViewer.ml

File: lib/HunkView.ml (hunk 15 above)
Comment (yokurang) [reply lib/HunkView.ml:122]: Use fold or something

File: lib/HunkView.mli (hunk 11 above)
Comment (yokurang) [reply lib/HunkView.mli:3]: We should define how to turn Patch.t Zipper -> Nottui.ui

Should not have Lwd stuff

The Lwd part should be extracted into InteractiveViewer.ml

File: lib/InteractiveViewer.mli (hunk 6 above)
Comment (yokurang) [reply lib/InteractiveViewer.mli:1]: Edit the comment

File: lib/InteractiveViewer.mli (hunk 4 above)
Comment (yokurang) [reply lib/InteractiveViewer.mli:21]: Just write intent of the function or important details that the function relies on to work 

If the function is obvious, we can only put a few words to keep the documentation minimal

File: lib/InteractiveViewer.ml [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+                 | `ASCII 'q', [] ->
+                     toggle_help_visibility ();
+                     `Handled
+                 | _ -> `Unhandled)
+               (W.string "Type 'q' to exit the help panel");
+        ]
+    else
+      W.vbox
+        [
+          Lwd.pure (OperationView.operation_info (Lwd.peek z_patches_var));
```
Comment (yokurang) [inline lib/InteractiveViewer.ml:57]: There is a bug where the changes are not reactive.. I think the solution lies somewhere in Lwd.map but after trying I can't seem to figure it out. Can I ask your advice on what I should do to resolve this? Thanks

File: lib/InteractiveViewer.ml [hunk 17]
```diff
@@ -36,6 +36,10 @@ let view (patches : Patch.t list) =
       Lwd.set curr_scroll_state { state with position = state.W.bound }
     else Lwd.set curr_scroll_state state
   in
+  let project_op (op : Patch.t Zipper.t -> Nottui.ui) : ui Lwd.t =
+    let$ z_patches = Lwd.get z_patches_var in
+    op z_patches
```
Comment (Julow) [inline lib/InteractiveViewer.ml:41]: I think you rewrote `Lwd.map`:
```
          Lwd.map (Lwd.get z_patches_var) ~f:OperationView.operation_info;
```

Comment (Julow) [reply lib/InteractiveViewer.ml:41]: To clarify: `let$` is an alias to `Lwd.map`. This can be written in a shorter way without introducing a local function.

`Lwd.get` "observes" a variable, it returns a `Lwd.t` that changes every time the variable is changed.

File: lib/InteractiveViewer.ml [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+  in
+  let curr_scroll_state = Lwd.var W.default_scroll_state in
+  let change_scroll_state _action state =
+    let off_screen = state.W.position > state.W.bound in
+    if off_screen then
+      Lwd.set curr_scroll_state { state with position = state.W.bound }
+    else Lwd.set curr_scroll_state state
+  in
+  let project_op (op : Patch.t Zipper.t -> Nottui.ui) : ui Lwd.t =
+    Lwd.map (Lwd.get z_patches_var) ~f:op
```
Comment (yokurang) [inline lib/InteractiveViewer.ml:40]: @Julow I rewrote it so more explicit and easier to follow

Comment (Julow) [reply lib/InteractiveViewer.ml:40]: I don't think this is easier to follow than inlining the `Lwd.map` directly where it's needed. I'd say introducing the `project_op` makes things less explicit.
We can keep it like that for now though.

File: lib/HelpView.mli [hunk 19]
```diff
@@ -0,0 +1,5 @@
+val help_panel : Nottui.ui
+(** [help_panel] is the help panel, containing the help text. *)
+
+val help_keyboard_area : Nottui.ui
+(** [help_keyboard_area] is the help keyboard area, containing the keyboard shortcut to leave the help panel view. *)
```
Comment (panglesd) [inline lib/HelpView.mli:5]: This is not used because the logic for it is in `InteractiveViewer.ml`

```suggestion
```

File: lib/HelpView.ml [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+      W.string ~attr:shortcut_attr
+        "p:   Move to the previous operation, if present";
+      W.string ~attr:shortcut_attr "t:   Toggle view mode";
+      W.string ~attr:shortcut_attr "l:   Toggle line numbers";
+    ]
+
+let help_keyboard_area =
+  Ui.keyboard_area
+    (function `ASCII 'q', [] -> `Handled | _ -> `Unhandled)
+    (W.string ~attr:Notty.A.(fg lightblue) "Type 'q' to exit the help panel")
```
Comment (panglesd) [inline lib/HelpView.ml:23]:
```suggestion
```

File: lib/HelpView.mli [hunk 21]
```diff
@@ -0,0 +1,5 @@
+val help_panel : Nottui.ui
+(** [help_panel] is the help panel, containing the help text. *)
+
+val help_keyboard_area : Nottui.ui
```
Comment (Julow) [inline lib/HelpView.mli:4]: `help_keyboard_area` is not used outside of this module.

File: lib/InteractiveViewer.ml (hunk 18 above)
Comment (panglesd) [reply lib/InteractiveViewer.ml:40]: I think it is because to:
- Call `Lwd.get` once to get an `Lwd.t` value that reacts on `z_patches_var` changes
- Inline this function on places where it is used.

**Review by Julow — APPROVED** [review]
This is a general improvement, let's merge !

File: lib/InteractiveViewer.ml (hunk 18 above)
Comment (yokurang) [reply lib/InteractiveViewer.ml]: Added
